src/int_id.h

```diff
@@ -128,4 +128,37 @@ struct hash< int_id<T> > {
 };
 } // namespace std
 
+/**
+ * A tiny fixed hashtable-based cache that speeds up `int_id::obj()` calls.
+ * Exploits the following facts:
+ * 1. same ids are usually clustered together (i.e. fields, furn, ter)
+ * 2. cache is tiny -> less CPU cache misses
+ * 3. if "obj" was stored in the cache, it's guaranteed to be valid
+ * Intended to be used locally in situations when `int_id::obj()` is called in a tight loop.
+ * Create the cache before the loop and then instead of `id.obj()` call `cache.obj(id)`.
+ * @tparam T type of the object of the corresponding int_id
+ */
+template<typename T>
+class int_id_cache
+{
+    private:
+        static const int LOCAL_OBJ_CACHE_SIZE = 16;
+        int ids[LOCAL_OBJ_CACHE_SIZE] = { -1 };
```

Comment:
This initializes the first entry to -1 and the remainder to 0; is that really what you wanted?

Reply:
Yes, only the first element can have a collision with zero value, e.g. `ids[0] == static_cast<int> (null_int_id)`.

Reply:
Probably worth a comment to that effect because it might look like a mistake to the next person too.